[Tahoe] Switching tabs in Safari sometimes results in blank tabs

smfr opened the pull request:
#### c33ca42eaaf5ef926e12f3f235467138f97ac0f1
<pre>
[Tahoe] Switching tabs in Safari sometimes results in blank tabs
<a href="https://bugs.webkit.org/show_bug.cgi?id=297630">https://bugs.webkit.org/show_bug.cgi?id=297630</a>
<a href="https://rdar.apple.com/158632267">rdar://158632267</a>

Reviewed by Tim Horton.

When switching tabs via the mouse in Safari, there&apos;s a hover preview that uses a
_WKThumbnailView, and this persists until after the switch, when it&apos;s animated out. The
thumbnail view is associated with a WKWebView, and when present, can intercept the layer
passed to `WebViewImpl::setAcceleratedCompositingRootLayer()`, which prevents that layer
from being parented in the main view, resulting in blank contents.

298652@main changed the timing of the `setAcceleratedCompositingRootLayer()` call such
that it now happens while the thumbnail view is present, resulting in blank tabs until
Safari unparents the thumbnail view.

Fix this in the case where _WKThumbnailView has already generated the snapshot by having
`WebViewImpl::updateThumbnailViewLayer()` return a value indicating whether it actually
stole the root layer (i.e. when needing to generate the snapshot); if not, we parent the
root layer in the main view even when the _WKThumbnailView is present.

* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::setAcceleratedCompositingRootLayer):
(WebKit::WebViewImpl::updateThumbnailViewLayer):

Canonical link: <a href="https://commits.webkit.org/298935@main">https://commits.webkit.org/298935@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9a3057ae6f5202156a6ec96a82ca7f6b04757abe

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/117225 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36893 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/27506 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/123323 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/69205 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/00189416-720a-4aa5-83f8-52e1a9269fa4) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/37591 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/45482 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/88965 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/43639 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e2c197f1-3851-4175-9384-4e032d9d6d15) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120158 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29916 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105110 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69458 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28974 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66997 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99313 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/23397 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/126445 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44122 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/33146 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97633 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/44478 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101337 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97424 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24809 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42768 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20703 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/40472 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43995 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/49654 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/43451 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46796 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45147 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/9a3057ae6f5202156a6ec96a82ca7f6b04757abe. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49625

smfr committed:
[Tahoe] Switching tabs in Safari sometimes results in blank tabs
https://bugs.webkit.org/show_bug.cgi?id=297630
rdar://158632267

Reviewed by Tim Horton.

When switching tabs via the mouse in Safari, there's a hover preview that uses a
_WKThumbnailView, and this persists until after the switch, when it's animated out. The
thumbnail view is associated with a WKWebView, and when present, can intercept the layer
passed to `WebViewImpl::setAcceleratedCompositingRootLayer()`, which prevents that layer
from being parented in the main view, resulting in blank contents.

298652@main changed the timing of the `setAcceleratedCompositingRootLayer()` call such
that it now happens while the thumbnail view is present, resulting in blank tabs until
Safari unparents the thumbnail view.

Fix this in the case where _WKThumbnailView has already generated the snapshot by having
`WebViewImpl::updateThumbnailViewLayer()` return a value indicating whether it actually
stole the root layer (i.e. when needing to generate the snapshot); if not, we parent the
root layer in the main view even when the _WKThumbnailView is present.

* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::setAcceleratedCompositingRootLayer):
(WebKit::WebViewImpl::updateThumbnailViewLayer):

Canonical link: https://commits.webkit.org/298935@main
